include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+	MSPI_DEVICE_CONFIG_WRITE_INSTR  = BIT(11),
+	MSPI_DEVICE_CONFIG_INSTR_LEN    = BIT(12),
+	MSPI_DEVICE_CONFIG_ADDR_LEN     = BIT(13),
+	MSPI_DEVICE_CONFIG_MEM_BOUND    = BIT(14),
+	MSPI_DEVICE_CONFIG_BREAK_TIME   = BIT(15),
+	MSPI_DEVICE_CONFIG_ALL          = BIT_MASK(16),
+};
+
+/**
+ * @brief MSPI XIP access permissions
+ */
+enum mspi_xip_permit {
+	MSPI_XIP_READ_WRITE     = 0,
+	MSPI_XIP_READ_ONLY      = 1,
+};
+
+/**
+ * @brief MSPI Configure API
+ * @defgroup mspi_configure_api MSPI Configure API
+ * @{
+ */
+
+/**
+ * @brief Stub for timing parameter
+ */
+enum mspi_timing_param {
+	MSPI_TIMING_PARAM_DUMMY
+};
+
+/**
+ * @brief Stub for struct timing_cfg
+ */
+struct mspi_timing_cfg {
+
+};
+
+/**
+ * @brief MSPI device ID
+ * The controller can identify its devices and determine whether the access is
+ * allowed in a multiple device scheme.
+ */
+struct mspi_dev_id {
+	/** @brief device gpio ce */
+	struct gpio_dt_spec     ce;
+	/** @brief device index on DT */
+	uint16_t                dev_idx;
+};
+
+/**
+ * @brief MSPI controller configuration
+ */
+struct mspi_cfg {
+	/** @brief mspi channel number */
+	uint8_t                 ui8MSPIChannel;
+	/** @brief Configure operaton mode */
+	enum mspi_op_mode       eOPMode;
+	/** @brief Configure duplex mode */
+	enum mspi_duplex        eDuplex;
+	/** @brief DQS support flag */
+	bool                    bDQS;
+	/** @brief Software managed multi-slave enable */
+	bool                    bSwMultiSlave;
+	/** @brief GPIO chip-select line (optional) */
+	struct gpio_dt_spec     *pCE;
+	/** @brief GPIO chip-select line numbers (optional) */
+	uint32_t                ui32CeGpioNum;
+	/** @brief Slave number from 0 to host controller slave limit. */
+	uint32_t                ui32SlaveNum;
+	/** @brief Maximum supported frequency in MHz */
+	uint32_t                ui32MaxFreq;
+	/** @brief Whether to re-initialize controller */
+	bool                    bReinit;
+};
+
+/**
+ * @brief MSPI DT information
+ */
+struct mspi_dt_spec {
+	/** @brief MSPI bus */
+	const struct device     *bus;
+	/** @brief MSPI hardware specific configuration */
+	struct mspi_cfg         config;
+};
+
+/**
+ * @brief MSPI controller device specific configuration
+ */
+struct mspi_dev_cfg {
+	/** @brief Configure CE0 or CE1 */
+	uint32_t                ui32CENum;
+	/** @brief Configure frequency */
+	uint32_t                ui32Freq;
+	/** @brief Configure I/O mode */
+	enum mspi_io_mode       eIOMode;
+	/** @brief Configure data rate */
+	enum mspi_data_rate     eDataRate;
+	/** @brief Configure clock polarity and phase*/
+	enum mspi_cpp_mode      eCPP;
+	/** @brief Configure transfer endian */
+	enum mspi_endian        eEndian;
+	/** @brief Configure chip enable polarity */
+	enum mspi_ce_polarity   eCEPolarity;
+	/** @brief Configure DQS mode */
+	bool                    bDQSEnable;
+	/** @brief Configure number of clock cycles between
+	 * addr and data in RX direction
+	 */
+	uint32_t                ui32RXDummy;
```

Comment:
I'm unclear why these values are configured once per device- different instructions may require different numbers of dummy cycles, right?

Also, what is the intended use of the Write instruction and Read Instruction in this structure? Are these intended to enable memory mapped access to the MSPI device?

Reply:
Have you looked at all of the code in context?
mspi_xfer_config specifically?

That explicitly implies that you can reconfigure certain things for each transfer. Since each MSPI controller will implement those differently, it's up to the chip manufacturer to add what the wish to add.

Items in mspi.h can be used either in the controller or in the device.

Reply:
One other question- unless I've missed it, I don't see a method for supporting enhanced XIP (IE no command code needed, simply send the address). Perhaps we should add a boolean flag to indicate to the controller that enhanced XIP mode is desired, and a way for the controller to indicate support for this?

Reply:
The write and read instructions in `mspi_dev_config` are device properties. They represent the most frequent used transfer commands throughout run-time. And yes, they are also for configuring XIP but falls into the device specific controller settings.

As for enhanced XIP or DMA, I don’t have devices and controllers that supports this feature and I’m unfamiliar with the flow. Maybe it is better to defer this from this PR, if it does not introduces conflict with the current proposal. I’d imagine it be added to both mspi_config and mspi_dev_config.

Reply:
Personally, I think we at least need to consider what XIP/memory mapped access looks like in this PR (although we don't have to implement it here). I want to avoid needing major API changes to implement XIP support.

At least on the NXP FLEXSPI, memory mapped access looks like so:
- the FLEXSPI has a programmable lookup table which stores command descriptors for various flash operations (IE send command, send address, read data, etc). The FLEXSPI can be instructed to manually execute any of these command descriptors via register access
- alternatively for memory mapped access, the FLEXSPI can have a READ and WRITE command index programmed for each device (up to 4 attached devices are supported, and they will be memory mapped sequentially in the FLEXSPI address space based on the chip select port used)
- The READ and WRITE command index refers to a specific index in the lookup table. The FLEXSPI can run multiple instructions for one READ or WRITE, since it executes commands until it encounters a special STOP command

To summarize, I think the current `mspi_dev_config` would work well for memory mapped access with the FLEXSPI. The only API I see missing is a way to get the address of a memory mapped device, so that the program can use that address as flash or RAM. We might want a way to read that address statically, since I can imagine users wanting to link data to a PSRAM region or something similar.

Reply:
I do find it curious that you had to implement a mspi_xip_get_address for the fact that the physical XIP start address is fixed for each of the hardware instance. What I did is to write the start address and size down in devicetree and get it via DT macros.  Also to properly link code, you'd have the XIP memory region defined.
<img width="537" alt="xip_region" src="https://github.com/zephyrproject-rtos/zephyr/assets/155429447/d7f3c089-dba0-4a2f-b912-e05a43030d0f">

Reply:
> I do find it curious that you had to implement a mspi_xip_get_address for the fact that the physical XIP start address is fixed for each of the hardware instance. What I did was to write the start address and size down in devicetree and get it via DT macros. Also to properly link code, you'd have the XIP memory region defined.

Yes, I agree. I think this should probably be moved to devicetree in this API, so that code can be linked to these regions. It's also worth noting that once multiple devices are attached to the FLEXSPI, the address region of each is no longer fixed. Once all devices are configured, their address regions will be sequential (devices would use individual mode here):

![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/d30bba77-11d3-499b-b2fd-6d0a289116a0)

However, if FLASH_A1 was not yet configured, the address space of FLASH_A2 would start at the beginning of the XIP region until FLASH_A1 was initialized. At runtime, this rarely makes a difference as all flash devices will be set up. However, it does affect things during init time. If we want to link code into these regions, we will need to implement some sort of function similar to `z_data_copy`: https://github.com/zephyrproject-rtos/zephyr/blob/23a03b7a0a494e27316fdd7fdb453a0409b99aa9/kernel/xip.c#L26 that runs after all MSPI memory mapped devices are initialized. Just something worth considering if we want complete XIP support

Reply:
Couldn't you determine the address location of each slave depending on its size and device index (from `struct mspi_dev_id`) and configure accordingly? If FLASH_A2 can not get to the correct address without FLASH_A1 being initialized, this seems rather couter intuitive. 
I agree that an general MSPI XIP region binding and corresponding support to move the data are needed. But lets think about it and I prefer to defer it to future enhancement, as it may not be a part of the API but a kernel enhancement.

Reply:
> If FLASH_A2 can not get to the correct address without FLASH_A1 being initialized, this seems rather couter intuitive

You'll have to take that up with NXP's hardware designers :). Essentially, if the size of flash A1 is set within the FLSHA1CR0 register, then the base address of flash A2 will be shifted by the value written to FLSHA1CR0 * 1024. We could theoretically write code for the FLEXSPI to initialize all FLSHAxCR0 registers at boot time, before each individual flash device initialized. However, this wouldn't add much value since the flash devices still couldn't be accessed until they were initialized.

Anyways, no reason to get into a tangent here- as you pointed out, this will likely need to be a future enhancement. I think the main thing we need at this stage is a way to read the base address for each MSPI device at build time, so that can be integrated with the linker and kernel in the future.

Reply:
I saw that the flexspi XIP start address and size has already been written to nxp soc dts.
```
flexspi: spi@134000 {
	reg = <0x40134000 0x1000>, <0x08000000 DT_SIZE_M(128)>;
};
```
But there is no xip regions. Maybe you can create partitions in a XIP region for a flexspi Or individual memory region to represent each device memory mapped area, since the device size and index are fixed during run-time. However we would not be able to use zephyr_code_relocate as `z_data_copy` is invoked before any device initialization.

Another way is to create partition under controller node, seems to related to https://github.com/zephyrproject-rtos/zephyr/issues/34167.

Reply:
> However we would not be able to use zephyr_code_relocate as z_data_copy is invoked before any device initialization.

See https://github.com/zephyrproject-rtos/zephyr/pull/68888 for a related problem on delayed init for relocated data. Tbc isf similar approach can be used for code relocation.